### DISCUSSION TITLE
Using tenv as a Go module?

### DISCUSSION THREAD

**Discussion by corrieriluca** [tofuutils/tenv#233] [Q&A] [answered]
Hi all,

I was wondering if `tenv` was suitable to be used as a Go module inside a Go codebase. We would like to use it as a replacement of Hashicorp's [hc-install](https://github.com/hashicorp/hc-install) module which only supports Terraform.

From what I understand, reading documentation and code, `tenv` does not have a stable and usable Go API yet. Could anyone confirm this?

Given the current state, would you recommend to use `tenv` as a CLI tool that we would call from our Go code?

Thank you in advance for your help!

Best regards,
Luca

**Comment by dvaumoron** [comment, marked as answer]
Hi @corrieriluca,

The Go API did not change often (although sometimes new feature or fix need it), however we haven't set compatibility goal on it.

Usage could look like :

```
conf, err := config.InitConfigFromEnv()
conf.InitDisplayer(false)
hclParser := hclparse.NewParser()
versionManager := builder.Builders[toolName](conf, hclParser)

err = versionManager.Install("latest") // can use semantic.LatestKey constant, other supported value : version, version contraints, known behavior
```

for more on VersionManager available method (see [go doc](https://pkg.go.dev/github.com/tofuutils/tenv/v3@v3.0.0/versionmanager)).

**Reply by corrieriluca** [reply]
Thank you for your response, I managed to use it properly 👍

**Comment by kvendingoldo** [comment]
@corrieriluca I think that you can use it as go library. We change the major version in case of changing anything major in the behavior.

**Reply by dvaumoron** [reply]
Maybe it could be interesting for this kind of use case to wrap the initialization part to make it simpler to call, any thoughts ?

**Reply by dvaumoron** [reply]
I continue to think about that, the proxy part could also benefit from a simplified API, those change could be made within a new package, a compatibily promise (linked to semantic versioning) could be done on this new package and versionmanager package (with ReleaseInfoRetriever interface for user wanting to use our code as a base to support more tools).

**Reply by kvendingoldo** [reply]
I think the separate package with compatibility promise will work like a charm. Let's create a new repo for that in tofuutils group

**Reply by dvaumoron** [reply]
Which benefit did you envision with a separate repository ? If we keep this in the same repository, that will sync thing automatically (we can still make a different module in the same repository, even if i see no real benefit for that either)

In term of visibility and usability, it seem better to have a specific documentation (moreover our Readme is already of a discouraging length), but it can be on the side of the current one

**Reply by kvendingoldo** [reply]
I don't insist :) Let's keep it here. Regarding the documentation, I gonna work on moving it to a simple web format like git book in the near future.

**Reply by dvaumoron** [reply]
Hi @corrieriluca, @kvendingoldo,

I started to work on [tenvlib](https://github.com/tofuutils/tenv/blob/feat/tenv-as-lib/versionmanager/lib/lib.go), any remark about the API ?

**Reply by kvendingoldo** [reply]
Wow! Great. I will respond to this discussion if I have any ideas.

**Reply by corrieriluca** [reply]
That's great! That could fit my use case, I implemented much the same thing on my side, it's better to be integrated into tenv directly 👍 

Just one question: what is the equivalent of `tenv <tool> detect`? The `Evaluate` function?

**Reply by dvaumoron** [reply]
I didn't report the equivalent of Detect yet 😅

**Reply by dvaumoron** [reply]
I think the ReadDefaultConstraint, SetConstraint and ResetConstraint will be useful too
